### ISSUE TITLE
Add support for @odata.bind

### ISSUE THREAD

**Issue by vojtechvit** [object/Simple.OData.Client#95]
Right now, when I try to use Simple OData Client to create an entity and to link its relationships at the same time using [@odata.bind syntax](http://docs.oasis-open.org/odata/odata/v4.0/errata02/os/complete/part1-protocol/odata-v4.0-errata02-os-part1-protocol-complete.html#_Toc406398327), the library throws an exception: `Simple.OData.Client.UnresolvableObjectException: No property or association found for inviter@odata.bind`.

## Sample code
``` c#
    var client = new ODataClient("http://localhost:15407/api/v1");
    var invitation = client
        .For("invitations")
        .Set(new Dictionary<string, object>
        {
            { "emailAddress", "my@email.com" },
            { "inviter@odata.bind", "users(79)" }
        })
        .InsertEntryAsync().Result;
```

**Comment by object** [comment]
Thank you for the suggestion. I will investigate how this can be added and come back to you.

**Comment by vojtechvit** [comment]
:+1:

**Comment by object** [comment]
Hi again, I checked how odata.bind works, but even though Simple.OData.Client doesn't support this syntax directly, doesn't it provide an equivalent functionality?

I mean, if you are using untyped syntax, you can refer to related entities either by their IDs or using navigation property names, see these examples from the library tests (https://github.com/object/Simple.OData.Client/blob/master/Simple.OData.Client.Tests.Net40/InsertTests.cs):

1.
var category = await _client
  .For("Categories")
  .Set(new { CategoryName = "Test3" })
  .InsertEntryAsync();
var product = await _client
  .For("Products")
  .Set(new { ProductName = "Test4", UnitPrice = 18m, CategoryID = category["CategoryID"] })
  .InsertEntryAsync();

2. 
var category = await _client
  .For("Categories")
  .Set(new { CategoryName = "Test5" })
  .InsertEntryAsync();
var product = await _client
  .For("Products")
  .Set(new { ProductName = "Test6", UnitPrice = 18m, Category = category })
  .InsertEntryAsync();

One of the goals of the library is to encapsulate OData conventional syntax, so as along as it provides equivalents that naturally fits .NET/C#, I'd rather use them.

**Comment by vojtechvit** [comment]
Hi Vagif,

Thanks for the answer. First an important note - in our company we want to use OData v4 as a standard protocol. Hence, some of our teams started creating endpoints that are OData v4 compliant, but we found out that client side tooling often doesn't support OData v4 features that are required in some our endpoints.. :-/

In the first example that you mentioned you set the category by setting a foreign key property, right? However, in most OData v4 models you don't have such properties. OData v4 does not even recognize the concept of Foreign Key Properties - there are only Navigation Properties.

In the second example you use a Navigation Property. However, what HTTP request payload will this produce (the last InsertEntryAsync)? Because it feels it will look something like this:
```
POST http://....
{
    'name': 'TestProduct1',
    'price': 18,
    'category': {
        'name': 'TestCategory1'
    }
}
```
... which, however is semantically different request from this:
```
POST http://....
{
    'name': 'TestProduct1',
    'price': 18,
    'category@odata.bind': "./categories('TestCategory1')"
}
```
The first request is, in OData v4, a deep insert [[Protocol]](http://docs.oasis-open.org/odata/odata/v4.0/errata02/os/complete/part1-protocol/odata-v4.0-errata02-os-part1-protocol-complete.html#_Toc406398328)[[JSON]](http://docs.oasis-open.org/odata/odata-json-format/v4.0/errata02/os/odata-json-format-v4.0-errata02-os-complete.html#_Toc403940637). You tell the service that you want a category created with that name along with the product.

The second request, however, says that you want to use an existing category as the navigation property value. And this is our case - our endpoint does not allow deep insert, but we want to allow the navigation property to be set on insert.

**Comment by vojtechvit** [comment]
I know that finding out how to represent this in static typed C# isn't straightforward... :-) Your library is, however, very nicely flexible thanks to the possibility to use dictionaries. Therefore, I'd add the support there (to make the example at the top of this page work) and only later perhaps something more advanced.. :-)

And just for you to know - Microsoft guys from Web API OData are actually facing the same challenge now - see https://github.com/OData/WebApi/issues/158 ;-D

**Comment by object** [comment]
Thank you for a great description of why you need this feature. I agree it's different from what I thought and what is currently supported by the library. I will think about how this can be added in a non-intrusive manner, so it fits in the current API syntax.

**Comment by object** [comment]
Hi again,

Actually I was a bit quick to agree that Simple.OData.Client is not capable of this feature. Apparently it does exactly this. One of its integration tests uses OData V4 to insert a product with an existing category:

            var category = await _client
                .For("Categories")
                .Set(CreateCategory(1005, "Test5"))
                .InsertEntryAsync();
            var product = await _client
                .For("Products")
                .Set(CreateProduct(1007, "Test6", category))
                .InsertEntryAsync();


And here's the generated payload for the second statement:

{
  "@odata.type":"#ODataDemo.Product","ID":1007,"Name":"Test6","Description":"Test1","Price":18.0,"Rating":1,"ReleaseDate":"2015-04-08T16:58:07.3126277+02:00","Categories@odata.bind":[
    "http://services.odata.org/V4/OData/(S(gj4yrh3yr0c4331antgquemh))/OData.svc/Categories(1005)"
  ]
}

Methods CreateProduct and CreateCategory simply create dictionaries of properties.

The bottom line: the library detects if you are creating a new related entity or referencing an existing one, and in the latter case it references it using odata.bind.

**Comment by vojtechvit** [comment]
Wow, you're right - it does work actually! :-)

``` c#
    var inviter = new { id = 79 };

    var invitation = client
        .For("invitations")
        .Set(new 
        {
            emailAddress = "my@email.com",
            inviter = inviter
        })
        .InsertEntryAsync().Result;
```

However, I wonder how does Simple OData Client find out what is the correct EntitySet (in this case "Users") and what is the key property (in this case "id")? Does it get this information from the $metadata endpoint? If so, when does it send the request to the Metadata Document? On the first (other) request just once? Or every time? And does it mean that Simple OData Client requires the Metadata Document to work?

Sorry, this is no longer related to the reported issue, but I'm curious.. :-)

**Comment by object** [comment]
Yes, Simple.OData.Client relies (and quite heavily) on a service metadata document that is retrieved prior to the first request to the service and cached in a dictionary of URL/Metadata pairs. So it knows more or less everything about the entity relationship and is able to build proper URIs and associations.

**Comment by vojtechvit** [comment]
Ok, thanks for info! ;-)

**Comment by poojai1214** [comment]
Hi,
We are still having issues linking lookup field while creating a record..We are on CRM V9. We need to link an existing account record to a Custom entity on create of Custom entity..
We have used code like below:
_cal_accountid_value = new { accountid = new Guid("xxxxxx-xxxx-xxxx-xxxxxxxxxxxx") }

I tried using "Id" instead of accountid..With the above code i get error message about Anonoymous System.Object type conversion error..

Any help will be really appreciated..

**Comment by davidnmbond** [comment]
We are suffering the same here.  CRM v9 (Dynamics 365) requires the format:
`
{
    "customerid_account@odata.bind": "/account(00000000-0000-0000-0000-0000-000000000000)"
}`

Looks like we're unable to use this library for CRM / Dynamics?

**Comment by mVermaat** [comment]
> We are suffering the same here. CRM v9 (Dynamics 365) requires the format: ` { "customerid_account@odata.bind": "/account(00000000-0000-0000-0000-0000-000000000000)" }`
> 
> ...but Fiddler shows that this is converted in the POST payload to... ` { "_accountid_value": "00000000-0000-0000-0000-0000-000000000000" }`
> 
> Looks like we're unable to use this library for CRM / Dynamics?

As I was looking for a solution to this issue and this github issue appeared on top, here's the solution to this problem:

````
            return new Dictionary<string, object>()
            {
                { "name", "anything" },
                { "customerid_account", new Dictionary<string, object>()
                    {
                        { "accountid", Guid.Parse("b67cb93e-a54e-ec11-8c62-6045bd88dd9d") },
                    }
                },
            };
````

You can also do the same with bound models:

````
[JsonPropertyName("customerid_account")]
public Account Account { get; set; }
````

````
    public class Account
    {
        [JsonPropertyName("accountid")]
        public Guid Id { get; set; }
    }
````
